zerver/lib/export.py

```diff
@@ -0,0 +1,653 @@
+            if sub.user_profile.realm != realm:
+                # In almost every case the other realm will be zulip.com
+                unsafe_huddle_recipient_ids.add(sub.recipient_id)
+
+        # Now filter down to just those huddles that are entirely within the realm.
+        huddle_subs = [sub for sub in realm_huddle_subs if sub.recipient_id not in unsafe_huddle_recipient_ids]
+        huddle_recipient_ids = set(sub.recipient_id for sub in huddle_subs)
+        huddle_ids = set(sub.recipient.type_id for sub in huddle_subs)
+
+        huddle_subscription_dicts = [model_to_dict(x) for x in huddle_subs]
+        huddle_recipients = [model_to_dict(r) for r in Recipient.objects.filter(id__in=huddle_recipient_ids)]
+        response['zerver_huddle'] = [model_to_dict(h) for h in Huddle.objects.filter(id__in=huddle_ids)]
+    else:
+        huddle_recipients = []
+        huddle_subscription_dicts = []
+
+    response["zerver_recipient"] = user_recipients + stream_recipients + huddle_recipients
+    response["zerver_subscription"] = user_subscription_dicts + stream_subscription_dicts + huddle_subscription_dicts
+
+    response["zerver_attachment"] = [model_to_dict(x) for x in
+                                     Attachment.objects.filter(realm=realm)]
+    floatify_datetime_fields(response, 'zerver_attachment', 'create_time')
+
+def fetch_usermessages(realm, message_ids, user_profile_ids, message_filename):
+    # UserMessage export security rule: You can export UserMessages
+    # for the messages you exported for the users in your realm.
+    #
+    # TODO: Optimize this query to do the same thing more efficiently
+    user_message_query = UserMessage.objects.filter(user_profile__realm=realm,
+                                                    message_id__in=message_ids)
+    user_message_chunk = [model_to_dict(um) for um in user_message_query if
+                          um.user_profile_id in user_profile_ids]
+    logging.info("Fetched UserMessages for %s" % (message_filename,))
+    return user_message_chunk
+
+def export_usermessages_batch(input_path, output_path):
+    with open(input_path, "r") as input_file:
+        output = ujson.loads(input_file.read())
+    message_ids = [item['id'] for item in output['zerver_message']]
+    user_profile_ids = output['zerver_userprofile_ids']
+    del output['zerver_userprofile_ids']
+    realm = Realm.objects.get(id=output['realm_id'])
+    del output['realm_id']
+    output['zerver_usermessage'] = fetch_usermessages(realm, message_ids, user_profile_ids, output_path)
+    write_message_export(output_path, output)
+    os.unlink(input_path)
+
+def write_message_export(message_filename, output):
+    with open(message_filename, "w") as f:
+        f.write(ujson.dumps(output, indent=4))
+    logging.info("Dumped to %s" % (message_filename,))
+
+def export_messages(realm, user_profile_ids, recipient_ids,
+                    chunk_size=1000, output_dir=None,
+                    threads=0):
```

Comment:
maybe instead of passing in `threads`, pass in `write_partials` as a boolean, or call it `num_threads` (this function doesn't use the actual number of threads in any meaningful way)